src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+        this one.
+        */
+        if (ste.blockType == ClassBlock) {
+            /* Pass down known globals */
+            append(newglobal, global);
+
+            /* Pass down previously bound symbols */
+            if (bound) {
+                append(newbound, bound);
+            }
+        }
+
+        for (const [name, flags] of Object.entries(ste.symbols)) {
+            this.analyze_name(ste, scopes, name, flags, bound, local, free, global);
+        }
+
+        /* Populate global and bound sets to be passed to children. */
+        if (ste.blockType !== ClassBlock) {
+            /* Add function locals to bound set */
+            if (ste.blockType === FunctionBlock) {
+                append(newbound, local);
+            }
+            /* Pass down previously bound symbols */
+            if (bound) {
+                append(newbound, bound);
+            }
+            /* Pass down known globals */
+            append(newglobal, global);
+        } else {
+            /* Special-case __class__ */
+            newbound.add("__class__");
+        }
+
+        /* Recursively call analyze_child_block() on each child block.
+        newbound, newglobal now contain the names visible in
+        nested blocks.  The free variables in the children will
+        be collected in allfree.
+        */
+        const allfree = new globalThis.Set<string>();
+        for (const entry of ste.children) {
+            this.analyze_child_block(entry, newbound, newfree, newglobal, allfree);
+            if (entry.hasFree || entry.childHasFree) {
+                ste.childHasFree = true;
+            }
+        }
+
+        append(newfree, allfree);
+
+        /* Check if any local variables must be converted to cell variables */
+        if (ste.blockType === FunctionBlock) {
+            this.analyze_cells(scopes, newfree);
+        } else if (ste.blockType === ClassBlock) {
+            ste.drop_class_free(newfree);
+        }
+
+        /* Records the results of the analysis in the symbol table entry */
+        ste.update_symbols(ste.symbols, scopes, bound, newfree, ste.blockType === ClassBlock);
+
+        append(free, newfree);
+    }
+}
+
+export function BuildSymbolTable(mod: mod, filename: string, future: never): SymbolTable {
+    const st = new SymbolTable(filename, future);
+    /* Make the initial symbol information gathering pass */
+    st.enterBlock("top", ModuleBlock, mod, 0, 0);
+    st.top = st.cur;
+    switch (mod._kind) {
+        case ASTKind.Module: {
+            st.SEQ(st.visitStmt, (mod as Module).body);
+            break;
+        }
+        case ASTKind.Expression: {
+            st.visitExpr((mod as Expression).body);
+            break;
+        }
+        case ASTKind.Interactive: {
+            st.SEQ(st.visitStmt, (mod as Interactive).body);
+            break;
+        }
+        case ASTKind.FunctionType:
+            throw new Error("this compiler does not handle FunctionTypes");
```

Comment:
ah that seems sad 😢 .

Reply:
https://github.com/python/cpython/blob/3.9/Python/symtable.c#L320-L323